GatekeeperStudyHall/Assets/Scripts/DiceRoll.cs

```diff
@@ -0,0 +1,203 @@
+
+    [Space]
+    [SerializeField] Sprite[] sprites; // the 6 dice faces
+
+
+    int roll = -1;
+    public UnityEvent<int> endEvent; // functions to be called after the dice is done rolling
+
+
+    void Start() {
+        if (sprites.Length != 6) {
+            Debug.LogError($"There should be 6 sprites in DiceRoll array (actual: {sprites.Length})");
+        }
+
+        // temporary debug message
+        endEvent.AddListener(x => Debug.Log($"You rolled a {roll}."));
+
+        spriteRenderer = GetComponentInChildren<SpriteRenderer>();
+        rb = GetComponentInChildren<Rigidbody2D>();
+        barrier.SetActive(false);
+    }
+
+    void Update() {
+        if (isHeld) {
+            UpdateHeldDice();
+        } 
+
+        else if (isSliding) {
+            UpdateSlidingDice();
+        }
+    }
+
+
+    // called whenever the dice is clicked on
+    public void MouseDownFunc() {
+        // we will likely have to change the condition here to account for
+        // all the times you are and aren't allowed to pick up the dice
+        if (!isSliding) {
+            isHeld = true;
+            shakeTimer = shakeInterval;
+        }
+    }
+
+
+    private void UpdateHeldDice() {
+
+        // move the dice base to the mouse's x and y position
+        Vector3 mousePosition = Input.mousePosition;
+        mousePosition = Camera.main.ScreenToWorldPoint(mousePosition);
+        mousePosition.z = transform.position.z;
+        transform.position = mousePosition;
+
+        // shake the dice once every `shakeInterval` seconds
+        if (shakeTimer >= shakeInterval) {
+
+            // change the face to show a random number
+            spriteRenderer.sprite = sprites[Random.Range(0, 6)];
+
+            // get a random point `shakeRadius` units away from the mouse,
+            // then start moving the dice sprite towards it
+            float unitCircleAngle = Random.Range(0f, 360f);
+            Vector3 unitCircleVec = new Vector3(Mathf.Cos(unitCircleAngle), Mathf.Sin(unitCircleAngle), 0);
+            Vector3 targetPos = unitCircleVec * shakeRadius;
+            Vector3 newVel = targetPos - rb.transform.localPosition;
+            newVel.z = 0;
+            rb.velocity = newVel.normalized * shakeSpeed;
+
+            // randomize angular velocity
+            float angVel = Random.Range(minAngularVel, maxAngularVel);
+            if (Random.value <= 0.5f) {
+                angVel = -angVel;
+            }
+            rb.angularVelocity = angVel;
+
+            // un-tick the timer
+            shakeTimer -= shakeInterval;
+        }
+
+        // tick the timer
+        shakeTimer += Time.deltaTime;
+    }
+
+
+    // called whenever the dice stops being clicked on
+    public void MouseUpFunc() {
+        if (!isHeld) {
+            return;
+        }
+
+        isHeld = false;
+        isSliding = true;
+        slideTimer = 0;
+
+        // determines the roll that this dice will get
+        roll = Random.Range(1, 7);
```

Comment:
I'd prefer you stay consistent and use `Random.Range(0, 6)`

Reply:
maybe rename `roll` to `rollIndex` to make it obvious you're using roll as the sprites index accessor

Reply:
`roll` is going to be used for more than just determining the sprite displayed on the dice in the future. I think its more clear for the `roll` variable to store values between 1 - 6 than 0 - 5. 
The line:
`spriteRenderer.sprite = sprites[Random.Range(0, 6)];` 
uses the range 0 - 5 because `sprites` is an array and needs to be indexed from 0 - 5.